### PR TITLE
Retry OSX support

### LINKED ISSUE #2: Build for OSX
I don't know what it would take but I would be interested in having an OSX package for ray.

### PR DESCRIPTION
**Pull request by h-vetinari**
Follow-up from #77 (which already did most of the groundwork), as well as #21

Fixes #2

Also cleans up some dependency cruft.

### CODE REVIEW

**Comment by h-vetinari** [comment]
@mattip
Did you ever have a passing OSX build in #77. This fails with
```
Execution platform: @local_config_platform//:host
make[3]: external/local_config_cc/cc_wrapper.sh: No such file or directory
```
presumably because our patched rule for building redis is not yet compatible with the OSX compiler stack (i.e. will probably need to use the LLVM versions and make sure the cc_wrapper is found):
```
  /bin/bash -c 'source external/bazel_tools/tools/genrule/genrule-setup.sh; 
        export CC=external/local_config_cc/cc_wrapper.sh
        export CFLAGS=
        export AR=${CC/gnu-cc/gnu-ar}
        export NM=${CC/gnu-cc/gnu-nm}
        export RANLIB=${CC/gnu-cc/gnu-ranlib}
        tmpdir="redis.tmp"
        p=external/com_github_antirez_redis/Makefile
        cp -p -L -R -- "${p%/*}" "${tmpdir}"
        chmod +x "${tmpdir}"/deps/jemalloc/configure
        parallel="$(getconf _NPROCESSORS_ONLN || echo 1)"
        make -s -C "${tmpdir}" -j"${parallel}" V=0 CFLAGS="${CFLAGS-} -DLUA_USE_MKSTEMP -Wno-pragmas -Wno-empty-body" AR="${AR}" RANLIB="${RANLIB}"
        mv "${tmpdir}"/src/redis-server bazel-out/darwin-opt/bin/external/com_github_antirez_redis/redis-server
        chmod +x bazel-out/darwin-opt/bin/external/com_github_antirez_redis/redis-server
        mv "${tmpdir}"/src/redis-cli bazel-out/darwin-opt/bin/external/com_github_antirez_redis/redis-cli
        chmod +x bazel-out/darwin-opt/bin/external/com_github_antirez_redis/redis-cli
        rm -r -f -- "${tmpdir}"
```

**Comment by mattip** [comment]
> Did you ever have a passing OSX build ...

No. I would prefer to move forward with #78 first, then return to this

**Comment by h-vetinari** [comment]
> > Did you ever have a passing OSX build ...
> 
> No. I would prefer to move forward with #78 first, then return to this

I picked the independently valuable changes from this PR into #81. Fine for me to do 2.1.0 first. I can rebase this PR once #78 lands.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -88,15 +77,15 @@ outputs:
...
         - protobuf >=3.15.3, <4.0.0
         - psutil
         - pyyaml
         - redis-py >=3.5.0  # [win]
         - requests
         - setproctitle =1.2.2
-        - typing_extensions  # [py<38]
         - virtualenv
+      run_constrained:
+        - __osx >={{ MACOSX_DEPLOYMENT_TARGET|default("10.15") }}  # [osx and x86_64]
```
Comment (vnlitvinov) [inline recipe/meta.yaml:88]: so this isn't needed on arm?

Comment (h-vetinari) [reply recipe/meta.yaml:88]: Arm only targets >=11.0 anyway

**Comment by h-vetinari** [comment]
@ngam, thanks for taking a look - would you like to collaborate on this PR?

**Comment by h-vetinari** [comment]
> I am going to try to help, but this is obviously going to be difficult (at least from the looks of it).

Yeah, not saying this will be easy, haha. Sent you an invite in any case. 🙃

**Comment by h-vetinari** [comment]
Currently based on #105, because why not

**Comment by h-vetinari** [comment]
So it looks like the main compilation succeed now, but we're running into
```
make[3]: external/local_config_cc/cc_wrapper.sh: No such file or directory
```
That sounds like a bazel thing? @ngam?

**Comment by mattip** [comment]
There are lots of these lines in the build log
```
Could not remove or rename \
   /Users/runner/miniforge3/conda-bld/ray-packages_1690346634210/_build_env/share/bazel/2966d1d2105c37b8a8f4040713b497b0/external/python3_9_x86_64-apple-darwin/lib/python3.9/site-packages/pip/_vendor/rich/tree.py.  
   Please remove this file manually (you may need to reboot to free file handles)
```

**Comment by h-vetinari** [comment]
> any history on the redis patch?

That was mainly @vnlitvinov's work. We [documented](https://github.com/conda-forge/ray-packages-feedstock/blob/main/ray-patching.md) how to generate it, but that doesn't cover the content of the redis patch itself.

**Comment by mattip** [comment]
I wonder if rebasing this off 2.6.2 would help.

**Comment by vnlitvinov** [comment]
> any history on the redis patch?

The point of those patches (IIRC there were multiple related to this topic, but maybe some were already upstreamed) is to make sure we're using the "correct" compiler. Building with `conda` is different from what Ray does at their own CI in the way that compilers/tools aren't named `cc`/`ar` etc, but are specified by a full name like `gcc-linux-whatever`, and are pointed to using environment variables.

So, when a build isn't expecting to use the variables like `$CC -c foo.c -o foo.bin` but instead runs `cc`, it fails in conda environment.

Looking at the quoted patch, I would guess that it was just a copy of the Linux one, and (probably) names of compiler and tools might be different on Darwin (i.e. maybe they aren't `gnu-whatever`).

**Comment by mattip** [comment]
Still getting this error, and lot of other build warnings.
```
external/local_config_cc/cc_wrapper.sh: No such file or directory
```

**Comment by h-vetinari** [comment]
Long-term, I'd like to avoid rebuilding redis completely (we have in in conda-forge), but to try that, we'll (well at least I) need to wait for ray to become compatible with bazel 6.

**Comment by vnlitvinov** [comment]
> Long-term, I'd like to avoid rebuilding redis completely

Ray "ideology" of vendoring most of the dependencies internally (which is kind of safe when you're running stuff in a random but production environment) conflicts with conda-forge "do not vendor stuff" pretty badly. Redis is just one of the cases (and I think initially Ray vendored Redis because they patched something in it).

**Comment by mattip** [comment]
Closing, #119 which replaces this was merged

**Pull request closed without merging**